In [1]:
import zipfile

with zipfile.ZipFile('archive.zip', 'r') as z:
    z.extractall('.')

print("Extracted!")

Extracted!


In [3]:
import pandas as pd
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv')
print(movies.shape)
print(credits.shape)

(4803, 20)
(4803, 4)


In [5]:
movies.head(2)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500


In [7]:
print(movies.columns.tolist())
print(credits.columns.tolist())

['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language', 'original_title', 'overview', 'popularity', 'production_companies', 'production_countries', 'release_date', 'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title', 'vote_average', 'vote_count']
['movie_id', 'title', 'cast', 'crew']


In [8]:

movies = movies[['id', 'title', 'overview', 'genres', 'keywords']]
credits = credits[['movie_id', 'title', 'cast', 'crew']]

print(movies.shape)
print(credits.shape)

(4803, 5)
(4803, 4)


In [9]:
movies.head()

,id,title,overview,genres,keywords
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...","[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na..."
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name..."
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...","[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,..."
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 818, ""name"": ""based on novel""}, {""id"":..."


In [12]:
credits.head()

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [21]:
movies = movies[['id', 'title', 'overview', 'genres', 'keywords']]
credits = credits[['movie_id', 'cast', 'crew']]

# Merge
movies = movies.merge(credits, left_on='id', right_on='movie_id')
movies.drop(columns=['movie_id'], inplace=True)

print(movies.columns.tolist())
print(movies.shape)

['id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew']
(4803, 7)


In [22]:
movies.head(2)

,id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


In [25]:
def convert_cast(obj):
    L = []
    counter = 0
    for i in ast.literal_eval(obj):
        if counter < 3:
            L.append(i['name'])
            counter += 1
        else:
            break
    return L


In [26]:
def fetch_director(obj):
    L = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [28]:
movies.head()

,id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [30]:
movies['cast'] = movies['cast'].apply(convert_cast)
movies['crew'] = movies['crew'].apply(fetch_director)

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]      [Sam Mendes]  


In [31]:
print(type(movies['cast'][0]))
print(movies['cast'][0])

<class 'list'>
['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver']


In [32]:
movies.head()

,id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


In [33]:
new_df = movies[['id', 'title']].copy()
new_df['tags'] = movies[['overview', 'genres', 'keywords', 'cast', 'crew']].astype(str).agg(' '.join, axis=1)
print(new_df['tags'][0])

In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. ['Action', 'Adventure', 'Fantasy', 'Science Fiction'] ['culture clash', 'future', 'space war', 'space colony', 'society', 'space travel', 'futuristic', 'romance', 'space', 'alien', 'tribe', 'alien planet', 'cgi', 'marine', 'soldier', 'battle', 'love affair', 'anti war', 'power relations', 'mind and soul', '3d'] ['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver'] ['James Cameron']


In [35]:
print(new_df.head())

       id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...  


In [36]:
print(new_df['tags'][0])

In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. ['Action', 'Adventure', 'Fantasy', 'Science Fiction'] ['culture clash', 'future', 'space war', 'space colony', 'society', 'space travel', 'futuristic', 'romance', 'space', 'alien', 'tribe', 'alien planet', 'cgi', 'marine', 'soldier', 'battle', 'love affair', 'anti war', 'power relations', 'mind and soul', '3d'] ['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver'] ['James Cameron']


In [37]:
movies.head()

,id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


# 🎬 Movie Recommendation System

## 📌 Project Overview
A Content-Based Filtering system that recommends similar movies
based on genres, keywords, cast, and director using the
TMDB 5000 Movies Dataset.

## 📂 Dataset
- `tmdb_5000_movies.csv` — Contains movie details such as
   genres, keywords, and overview
- `tmdb_5000_credits.csv` — Contains cast and crew information

## ⚙️ Workflow

### Step 1 — Import Libraries
Imported essential libraries including pandas, numpy,
ast, scikit-learn, and pickle.

### Step 2 — Data Loading
- Loaded both CSV files into pandas DataFrames
- Verified shapes: Movies (4803, 20) | Credits (4803, 4)
- Merged both DataFrames on movie ID

### Step 3 — Data Preprocessing
- Selected relevant columns: id, title, overview,
  genres, keywords, cast, crew
- Parsed JSON formatted columns to extract names
- Extracted top 3 actors from cast
- Extracted Director from crew
- Handled null values

### Step 4 — Feature Engineering
- Combined overview, genres, keywords, cast, and director
  into a single `tags` column
- Each movie is now represented as a single text string

## 🔜 Upcoming Steps
- Text Vectorization using CountVectorizer
- Similarity Calculation using Cosine Similarity
- Building Recommendation Function
- Streamlit App Development
- Deployment on Streamlit Cloud

In [38]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(max_features=5000, stop_words='english')
vectors = cv.fit_transform(new_df['tags']).toarray()

print(vectors.shape)

(4803, 5000)


In [43]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zooey', 'zucker', 'zwick'], dtype=object)

In [45]:
vocab = cv.get_feature_names_out()
print(vocab[:150])

['000' '007' '10' '100' '11' '12' '13' '14' '15' '16' '17' '17th' '18'
 '18th' '19' '1930s' '1940s' '1950s' '1960s' '1970s' '1980' '1980s' '19th'
 '20' '200' '20th' '21st' '24' '25' '30' '3d' '40' '50' '60' '60s' 'aaron'
 'abandoned' 'abducted' 'abigail' 'abilities' 'ability' 'able' 'aboard'
 'abraham' 'abroad' 'abuse' 'abusive' 'academy' 'accent' 'accept'
 'accepted' 'accepts' 'access' 'accident' 'accidental' 'accidentally'
 'accompanied' 'accomplish' 'account' 'accountant' 'accused' 'ace'
 'achieve' 'act' 'acting' 'action' 'actions' 'activist' 'activities'
 'activity' 'actor' 'actors' 'actress' 'acts' 'actual' 'actually' 'ad'
 'adam' 'adams' 'adaptation' 'adapted' 'addict' 'addicted' 'addiction'
 'adolescence' 'adopted' 'adoption' 'adopts' 'adrian' 'adrien' 'adult'
 'adultery' 'adulthood' 'adults' 'advantage' 'adventure' 'adventures'
 'advertising' 'advice' 'affair' 'affairs' 'affection' 'affections'
 'affleck' 'afghanistan' 'africa' 'african' 'aftercreditsstinger'
 'afterlife' 'afte

In [46]:
import nltk
from nltk.stem.porter import PorterStemmer

nltk.download('punkt')

ps = PorterStemmer()

def stem(text):
    return " ".join([ps.stem(word) for word in text.split()])

new_df['tags'] = new_df['tags'].apply(stem)

print(new_df['tags'][0])


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. ['action', 'adventure', 'fantasy', 'scienc fiction'] ['cultur clash', 'future', 'space war', 'space colony', 'society', 'space travel', 'futuristic', 'romance', 'space', 'alien', 'tribe', 'alien planet', 'cgi', 'marine', 'soldier', 'battle', 'love affair', 'anti war', 'power relations', 'mind and soul', '3d'] ['sam worthington', 'zoe saldana', 'sigourney weaver'] ['jame cameron']


In [47]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)
similarity[0]

array([1.        , 0.06885304, 0.04948717, ..., 0.03142697, 0.05410018,
       0.        ])

In [48]:
def recommend(movie):

    index = new_df[new_df['title'] == movie].index[0]


    similar_movies = sorted(list(enumerate(similarity[index])), reverse=True, key=lambda x: x[1])[1:6]

    for i in similar_movies:
        print(new_df.iloc[i[0]].title)

In [49]:
recommend('Avatar')

Aliens
Moonraker
Alien
Alien³
Silent Running


In [51]:
recommend('Batman Begins')

The Dark Knight
The Dark Knight Rises
Batman
Batman & Robin
Batman Returns


In [52]:
import pickle

pickle.dump(new_df, open('movies.pkl', 'wb'))
pickle.dump(similarity, open('similarity.pkl', 'wb'))

print("Files saved!")

Files saved!


## 🤖 Model Building

### Text Vectorization
- Converted movie tags into numbers using CountVectorizer
- Used top 5000 words only
- Removed common words like "the", "is", "and"
- Applied Stemming — "loving" becomes "love"

### Cosine Similarity
- Compared all 4803 movies with each other
- Higher score means more similar movies

### Recommendation Function
- Input: Movie name
- Output: Top 5 similar movies

### Files Saved
- `movies.pkl` — movie data
- `similarity.pkl` — similarity scores